Resolving some issues with UART and Clock

shayan0373n opened the pull request:
The first change is to avoid bypassing the crystal oscillator.
The second change is in UART RX pin mode where the reference manual recommended it to be FLOATING or PULLUP. In any case, PUSHPULL does not work.

Commit:
Resolving some issues with UART and Clock

feaser commented:
Hello shayan0373n,

Thank you for taking the effort of providing this feedback. 

After researching your proposed changes, I came to the following conclusion:

- The HSE bypass clock is needed on the Nucleo-F103RB board. On this board, there is no crystal oscillator directly connected to the microcontroller. Instead, the ST-Link 8 MHz MCO is used as the HSE clock source. Therefore the bypass is correct to turn off the XTAL loop output.

- The UART RX GPIO pin is configured for alternate mode, specifically UART RX mode in this case. The setting of GPIO_InitStruct.OutputType is therefore "don't care", because the GPIO pin is treated as an input. 

For these reasons I decided to not process your changes. Hopefully this doesn't discourage you from proposing future improvements.

Take care,

Frank

shayan0373n commented:
Dear Frank,

Thank you for your time and considerations.

Regarding the HSE bypass clock, you are right. I was porting the boot loader into my custom build STM32F103 board and wasn't aware of the fact you mentioned.

Regarding UART; I'm sorry, I have mistakenly edited the wrong line of the code. My proposed change should be like this:
``` C
#if (BOOT_COM_UART_ENABLE > 0)
  /* UART TX and RX GPIO pin configuration. */
  GPIO_InitStruct.Pin = LL_GPIO_PIN_2;
  GPIO_InitStruct.Mode = LL_GPIO_MODE_ALTERNATE;
  GPIO_InitStruct.Speed = LL_GPIO_SPEED_FREQ_HIGH; // !CHANGED!
  GPIO_InitStruct.OutputType = LL_GPIO_OUTPUT_PUSHPULL;
  LL_GPIO_Init(GPIOA, &GPIO_InitStruct);
  GPIO_InitStruct.Pin = LL_GPIO_PIN_3;
  GPIO_InitStruct.Mode = LL_GPIO_MODE_FLOATING; // !CHANGED!
  LL_GPIO_Init(GPIOA, &GPIO_InitStruct);
#endif
```
This is to comply with page 167 of RM0008 Rev 20 (STM32 F10x Reference Manual) Table 24.
As a matter of fact, as long as I didn't change the Pin Mode to FLOATING, the bootloader wouldn't work (It wouldn't receive any UART packages from the host).

Best Regards,
Shayan

![image](https://user-images.githubusercontent.com/17802921/59780866-efc40f80-92cf-11e9-853b-c165f448b209.png)

kruizer23 commented:
Hello Shyan,

After verifying that your proposed changes also work on the Nucleo-F103RB board, I committed them to the SVN repository over at SourceForge. This repository is automatically synced to GitHub daily, so the change will appear in the Github repository tomorrow.

Thanks for your feedback and participation in the OpenBLT project!

Take care,

Frank
